<a href="https://colab.research.google.com/github/vinalyan/pytonDS/blob/lession--3/pythonDS_LS3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Задание 1
1. Импортируйте библиотеки `pandas` и `numpy`.
2. Загрузите `"Boston House Prices dataset"` из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (`X_test, y_test`) с помощью функции `train_test_split` так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
3. Создайте модель линейной регрессии под названием `lr` с помощью класса `LinearRegression` из модуля `sklearn.linear_model`.
4. Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
5. Вычислите `R2` полученных предказаний с помощью `r2_score` из модуля `sklearn.metrics`.


In [24]:
import pandas as pd
import numpy as np

In [25]:
from sklearn.datasets import load_boston
X, y = load_boston(return_X_y=True)

In [26]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.33, random_state=42)

In [27]:
from sklearn.linear_model import LinearRegression
lr  = LinearRegression()
lr.fit(X_train,y_train)
y_pred = lr.predict(X_test)

In [28]:
from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.7261570836552481

# Задание 2
1. Создайте модель под названием `model` с помощью `RandomForestRegressor` из модуля `sklearn.ensemble`.
2. Сделайте агрумент `n_estimators` равным `1000`,
`max_depth` должен быть равен `12` и `random_state` сделайте равным `42`.
3. Обучите модель на тренировочных данных аналогично тому, как вы обучали модель `LinearRegression`, но при этом в метод `fit` вместо датафрейма `y_train` поставьте` y_train.values[:, 0]`,чтобы получить из датафрейма одномерный массив `Numpy`, так как для класса `RandomForestRegressor` в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
4. Сделайте предсказание на тестовых данных и посчитайте `R2`. Сравните с результатом из предыдущего задания.
5. Напишите в комментариях к коду, какая модель в данном случае работает лучше.



In [30]:
from sklearn.ensemble import RandomForestRegressor 
model = RandomForestRegressor(n_estimators = 1000, max_depth=12, random_state=42)

In [45]:
model.fit(X_train,y_train)
y_pred_tree = model.predict(X_test)

In [43]:
r2_score(y_test,y_pred_tree)

0.8668846861181482

Судя по коэффициенту детерминации лес работает поточнее. 

# *Задание 3
Вызовите документацию для класса `RandomForestRegressor`,
найдите информацию об атрибуте `feature_importances_`.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.


In [46]:
RandomForestRegressor?

In [51]:
model.feature_importances_

array([0.03396895, 0.00314105, 0.00626996, 0.00133789, 0.01189693,
       0.35801278, 0.01476964, 0.05994205, 0.004543  , 0.01247413,
       0.01902944, 0.01174753, 0.46286667])

Наиболее важными показателями являются показатели под номерами:
* 5 -  0.36
* 12 - 0.46

# *Задание 4
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.
Для этого датасета мы будем решать задачу классификации - будем определять,какие из транзакциции по кредитной карте являются мошенническими.Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.Мы будем вычислять AUC, то есть площадь под кривой ROC.
Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.
Загрузите датасет creditcard.csv и создайте датафрейм df.
С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована.Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
`pd.options.display.max_columns = 100`.
Просмотрите первые 10 строк датафрейма df.
Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.

Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.
Для поиска по сетке параметров задайте такие параметры:

`parameters = [{'n_estimators': [10, 15],'max_features': np.arange(3, 5),'max_depth': np.arange(4, 7)`

Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100),
param_grid=parameters,
scoring='roc_auc',
cv=3.
Обучите модель на тренировочном наборе данных (может занять несколько минут).
Просмотрите параметры лучшей модели с помощью атрибута best_params_.
Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументовмассивы y_test и y_pred_proba.
